### ISSUE TITLE
[Bug]: Agent Can't Access Files in Workspace

### ISSUE THREAD

**Issue by wTaylorBickelmann** [OpenDevin/OpenDevin#2408]
### Is there an existing issue for the same bug?

- [X] I have checked the troubleshooting document at https://opendevin.github.io/OpenDevin/modules/usage/troubleshooting
- [X] I have checked the existing issues.

### Describe the bug

I will assign a workspace when activating the docker container and I will see files I put there in the 'Code Editor' tab, but the agent is not able to find them. The agent will execute commands to find them (like `find` using the terminal and the Jupyter Ipython functionality, but still nothing. 

I have tried setting the workspace explicitly (OPENDEVIN_WORKSPACE=User/Will/Documents/workspace) and also using pwd (OPENDEVIN_WORKSPACE=$(pwd)/workspace)

I have tried using opendevin with ghcr.io/opendevin/opendevin:0.6 and ghcr.io/opendevin/opendevin:0.6.2 and the problem persists in either case.

![image](https://github.com/OpenDevin/OpenDevin/assets/57469687/d932f19f-23d6-407b-aacd-61cf2d7a4b97)

### Current OpenDevin version

```bash
0.6 & 0.6.2
```

### Installation and Configuration

```bash
OPENDEVIN_WORKSPACE=User/Will/Documents/workspace
docker run -it \
    --pull=always \
    -e SANDBOX_USER_ID=$(id -u) \
    -e PERSIST_SANDBOX="true" \
    -e SSH_PASSWORD="make something up here" \
    -e WORKSPACE_MOUNT_PATH=$OPENDEVIN_WORKSPACE \
    -v $OPENDEVIN_WORKSPACE:/opt/workspace_base \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 3000:3000 \
    --add-host host.docker.internal:host-gateway \
    --name opendevin-app-$(date +%Y%m%d%H%M%S) \
    ghcr.io/opendevin/opendevin:0.6
```

### Model and Agent

gpt4

### Operating System

MacOS

### Reproduction Steps

 - set workspace
 - load files into it
 - ask agent if it can find the files

**Comment by wTaylorBickelmann** [comment]
@SmartManoj the issue you linked looks similar, but on that ticket it says its just a WSL thing whereas I am using MacOS
Docker:  4.30.0

**Comment by mamoodi** [comment]
Yep I can reproduce this. MacOS. OpenDevin 0.6.
I have my $OPENDEVIN_WORKSPACE setup but the UI is showing something different.
![Screenshot 2024-06-12 at 8 33 33 PM](https://github.com/OpenDevin/OpenDevin/assets/169627366/c3168efe-5d2e-4c51-961a-552f8569738a)

**Comment by mamoodi** [comment]
@wTaylorBickelmann there is definitely an issue here. In your docker command can you set:
-e PERSIST_SANDBOX="false" instead of "true"

And see if the directory lines up with what OpenDevin sees?

**Comment by SmartManoj** [comment]
@wTaylorBickelmann Is the workspace directory changed?

**Comment by tobitege** [comment]
> Yep I can reproduce this. MacOS. OpenDevin 0.6. I have my $OPENDEVIN_WORKSPACE setup but the UI is showing something different.

The `OPENDEVIN_WORKSPACE` is an error in the docs, I think. There's no reference to it in the codebase, could be a remnant from earlier versions? I think it should be `WORKSPACE_BASE`

**Comment by SmartManoj** [comment]
> The `OPENDEVIN_WORKSPACE` is an error in the docs, I think.

It is just a variable.

> ```shell
>  -e WORKSPACE_MOUNT_PATH=$OPENDEVIN_WORKSPACE \
>  -v $OPENDEVIN_WORKSPACE:/opt/workspace_base \
> ```

Passed here.

**Comment by wTaylorBickelmann** [comment]
@mamoodi 
Changing that parameter seemed to work, as it can now see and access the files I put in the workspace. Thanks! 

![image](https://github.com/OpenDevin/OpenDevin/assets/57469687/0381ac9f-685a-465b-a265-5602f004c588)

**Comment by SmartManoj** [comment]
> @wTaylorBickelmann Is the workspace directory changed?

Previously, did you run in a different folder?

**Comment by wTaylorBickelmann** [comment]
No I used the same command but just changed that parameter

**Comment by neubig** [comment]
OpenDevin fixed the documentation in #2418 !

**Comment by mamoodi** [comment]
@neubig can you please clarify if your PR means that the OpenDevin Workspace can no longer be set and that the functionality has been removed?

**Comment by tobitege** [comment]
> @neubig can you please clarify if your PR means that the OpenDevin Workspace can no longer be set and that the functionality has been removed?

The "OPENDEVIN_WORKSPACE" name was a dummy and of no consequence. It was just to pass values into the docker statement as a temp variable. Only the WORKSPACE_BASE (or workspace_base in config) are the correct names.

**Comment by mamoodi** [comment]
@tobitege interesting. I'm having the same problem with WORKSPACE_BASE... 😄 
I set: WORKSPACE_BASE

Run the docker command as the README states and the OpenDevin workspace is not what I set it to.

**Comment by enyst** [comment]
@mamoodi could you perhaps try this branch to see if it fixes it? https://github.com/OpenDevin/OpenDevin/pull/2413

I assume you experience it still with `persist` set to true, not with `false`. I would clean up old docker images before either way.

**Comment by SmartManoj** [comment]
> I would clean up old docker images before either way.

only if not needed?

**Comment by mamoodi** [comment]
@enyst tried the branch and it's the same issue. What I did:
* Cloned the repo
* Ran **make setup-config** and set my workspace with persist_sandbox=true
* Ran **make run**
* Everything working and workspace is correct
* Stop OpenDevin. Modify **workspace_base** to a different workspace and keep persist_sandbox=true
* Ran **make run**

The folder it is using is not what is in my config. It's what was there before.
```
[core]
workspace_base="/Users/mahmoudwork/Desktop/workspaces/workspace_second"
ssh_password="a"
persist_sandbox=true
```

Then for fun, I tried setting `persist_sandbox=false` and even deleting my sandbox container and it still wasn't able to pick up my new workspace. So I don't believe the issue is fixed in that branch.

**Comment by SmartManoj** [comment]
In your docker desktop, how many containers were created?

**Comment by mamoodi** [comment]
> In your docker desktop, how many containers were created?

You will have to be more specific on when. For the first part:
```
Cloned the mentioned repo and branch
Ran make setup-config and set my workspace with persist_sandbox=true
Ran make run
Everything working and workspace is correct
Stop OpenDevin. Modify workspace_base to a different workspace and keep persist_sandbox=true
Ran make run
```

Only one container was there the entire time.

**Comment by SmartManoj** [comment]
> Only one container was there the entire time.

the container name? (The workspace name will be appended to the sandbox container)

**Comment by enyst** [comment]
> Does the 'persist_sandbox' even work in the config.toml? Because the variable would at least work in the docker container command and would pick up the new workspace.

Yes, it should work the same way as the env variable does. But... that makes me wonder. When running with `make build / make run`, what we have is: env vars are picked up in config and have priority, then config.toml values, then it falls back to our hardcoded defaults. Just a shot in the dark: did you have some env vars defined where you ran? They would override the .toml.

**Comment by mamoodi** [comment]
CC: @enyst and @SmartManoj 
Thanks for the help. The issue I ran into was I had WORKSPACE_BASE setup and that's why it wasn't working. I retested this issue with that PR and it seems to work. I'll think about how we can document it better. Thanks again

**Comment by enyst** [comment]
I mostly replicated the issue described by @mamoodi, using `main` branch, on Mac, building with `make build / make run` as in the post. Except for the UI part. I mean, there are two things here, aside from the toml:

1. the UI shows the old or new workspace. It shows the new workspace in my tests.
2. the agent is actually unable to use the new workspace, it sees the old workspace. This is the result I got.

Re: 1. Why the UI worked fine on my machine: I tracked it down to this https://github.com/OpenDevin/OpenDevin/blob/main/opendevin/runtime/server/runtime.py#L36, it correctly initialized the filestore with the new workspace, both when starting a new session and apparently when loading an old session. This is the source of the files displayed in the UI, which the UI requests with /api/list-files.

What could be the cause of displaying the old files in UI:
- is there any caching going on frontend? I didn't notice that, the UI always requests files
- is there some scenario in which loading an old session doesn't reinitialize the file store as above?
- was it an old browser tab and it didn't get to refresh the files when re-starting the session? I did re-use browser tabs and didn't see it, but... maybe?
- something else?

Re: 2. As for the cause of the agent seeing the old workspace, I think @SmartManoj tracked that down to the mount of the workspace directory. We create a container with two mounted volumes, one of them is mapping the workspace on the host machine to the /workspace in the sandbox (the other is a cache). The persisted container has that old volume, and that's what the agent sees.

If that is so, the PR should work, but my docker wasn't in a very cooperating mood today and I didn't manage to see it work. Also, this bug is not limited to Mac.

**Comment by mamoodi** [comment]
In the README we removed the PERSIST_SANBOX=true for now. Going to close this up.

**Comment by eliaspfeffer** [comment]
TL;DR: 

Had the same issue, this worked for me (added `  -e PERSIST_SANDBOX=true \` and added `-v ~/Documents/Github:/workspace \` )

```
docker run -it --rm --pull=always \
  -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:0.39-nikolaik \
  -e PERSIST_SANDBOX=true \
  -e LLM_API_KEY="ollama" \
  -e LLM_BASE_URL="http://host.docker.internal:11434" \
  -e LLM_OLLAMA_BASE_URL="http://host.docker.internal:11434" \
  -v /var/run/docker.sock:/var/run/docker.sock \
  -v ~/.openhands-state:/.openhands-state \
  -v ~/Documents/Github:/workspace \
  -p 3000:3000 \
  --add-host host.docker.internal:host-gateway \
  --name openhands-app \
  docker.all-hands.dev/all-hands-ai/openhands:0.39
```
